kolibri/plugins/h5p_viewer/assets/src/views/Html5AppRendererIndex.vue

```diff
@@ -78,7 +79,7 @@
     },
     methods: {
       recordProgress() {
-        const totalTime = now() - this.startTime;
```

Comment:
This is actually the correct way for a renderer to calculate the time spent - we should be using this in the other renderers, not using vuex getters.

Reply:
This calculates the time spent in the current session, and it's fine for that purpose.
However this time was being used to calculate the completion percentage and it's the reason why users are complaining of spending 50 minutes in the resource without being able to complete it.
If the user reads a pdf or see some html5 content in different days because he does not complete the whole time in a row, using totalTime will make the user unable to get the completion percentage if it's restarted everytime he begins because the previous spent time is not taken into account to calculate the completion percentage.

For example, in a pdf with 40 pages, completion time will be 40 x 30 = 1200 seconds. If the user reads 30 pages spending 15 minutes, and returns the next day to continue, all the previous time is discarded if we don't use the previous time spent (avaialable in the vuex getter) to calculate the percentage, so he might be reading the document forever without getting the completion time if he never spends those 1200 seconds (20 min) in a single session.

Reply:
Right - that's why we should be using `addProgress` not `updateProgress` - we should be adding to the total progress achieved from the progress in this session. If `addProgress` is not doing that properly, then that's a bug in `addProgress`.

If I read for 2 minutes on day one, and that gives me 12% completion.
On day two I read for 3 minutes and that gives me 18% completion.

If we add the progress, it should give a total of 30% completion, without any need to refer to to the previous day's progress.

Reply:
To use `addProgress`, we need to change the renderers to start passing the percent of progress since the last `addProgress` call was made. 

Right now, most renderers are sending the progress since session start instead, so when `addProgress` is used for updates, it leads to updates like this:

progress += 2%
progress += 4%
progress += 6%
progress += 8%
progress += 10%

So after five progress updates, instead of progress being the expected 10%, it is actually 30%.

Reply:
Right - because most of the renderers were originally using `updateProgress` so were not doing the incremental changes, so we need to update the implementation to handle that.

Reply:
(or at least they were coded that way - I think it may have just been totally wrong originally)

Reply:
Originally, there was only `updateProgress`, and some renderers (mainly video and audio) used it to add to progress, while others used it like a setter. 

I suspect it turned out this way because people are naturally inclined to code progress as a percent of completion rather than calculating a differential, and the term `update` usually implies setting rather than incrementing or otherwise combining the old and new value.

To address this, I created `addProgress` as an attempt to support both approaches and clarify the differences between them:

https://github.com/learningequality/kolibri/commit/95b6783abfbe919cf5e0a70b9491e1c84c7f8730

I think there are two separate issues here:

1) fixing the progress calculation bug, and
2) refactoring renderers to work in non-content contexts (I'm not familiar with this, and since many renderers use `sessionTimeSpent` getter, it appears to be a long-standing issue.)

Are you proposing that (2) should be fixed as part of this work? Otherwise, I would think the next action would be to create an issue for (2) and not forbid use of state getters in this PR as many renderers are already doing this.

Reply:
Yeah, I'd be fine with just fixing 1) - I just don't want the fix to add to 2) in the process if avoidable :)

Reply:
ok, so back to the first idea this PR had: 
- Inside a session we add progresss not update to avoid resetting previous session time
- The addProgress function in the store should be changed to avoid accumulating percentages, as Kevin commented previously currently, the renderers are sending the whole session percentage, causing a wng calculation.

But  taking into account that most of the renderers are doing it incorrectly, probably since ever, they all have to be modified.